File: whatsapp_notifications/whatsapp_notifications/page/whatsapp_schedule_monitor/whatsapp_schedule_monitor.py
"""
WhatsApp Schedule Monitor - Server-side data for the monitoring page
"""
import frappe
from frappe.utils import today, getdate, now_datetime, get_datetime, add_days, nowdate


@frappe.whitelist()
def get_schedule_data():
    """
    Return all scheduled rule data for the monitor page:
    - Birthday rules with next run, today's matches, last run info
    - Days Before/After notification rules with target date, matches, sent count
    - Recent activity from scheduled rules
    - Recent bulk sends (Envio em Massa WhatsApp)
    """
    return {
        "birthday_rules": _get_birthday_rules(),
        "date_event_rules": _get_date_event_rules(),
        "recent_activity": _get_recent_activity(),
        "bulk_sends": get_bulk_sends(limit=15),
        "now": str(now_datetime()),
        "today": today(),
    }


# ---------------------------------------------------------------------------
# Birthday Rules
# ---------------------------------------------------------------------------

def _get_birthday_rules():
    rules = frappe.get_all(
        "WhatsApp Birthday Rule",
        fields=[
            "name", "document_type", "birthdate_field", "send_time", "days_before",
            "last_run", "last_count", "last_status", "last_error", "enabled",
            "send_to_person", "send_to_group", "send_to_additional"
        ],
        order_by="enabled desc, name asc"
    )

    today_str = today()
    now = now_datetime()

    for rule in rules:
        rule["was_run_today"] = (
            bool(rule.get("last_run")) and
            getdate(rule["last_run"]) == getdate(today_str)
        )
        rule["next_run"] = _calc_birthday_next_run(rule, today_str, now)
        rule["matches_today"] = _count_birthday_matches(rule, today_str)
        rule["send_channels"] = _birthday_channels(rule)
        # Clean up internal fields not needed in UI
        rule.pop("birthdate_field", None)

    return rules


def _calc_birthday_next_run(rule, today_str, now):
    if not rule.get("enabled"):
        return None
    if not rule.get("send_time"):
        return None

    send_time_dt = get_datetime("{} {}".format(today_str, rule["send_time"]))

    if rule.get("was_run_today"):
        tomorrow = add_days(today_str, 1)
        return "{} {}".format(tomorrow, rule["send_time"])
    elif now < send_time_dt:
        return "{} {}".format(today_str, rule["send_time"])
    else:
        # send_time passed today but rule didn't run yet
        return "overdue"


def _count_birthday_matches(rule, today_str):
    if not rule.get("document_type") or not rule.get("birthdate_field"):
        return 0
    try:
        target = getdate(add_days(today_str, rule.get("days_before") or 0))
        result = frappe.db.sql(
            """SELECT COUNT(*) FROM `tab{doctype}`
               WHERE MONTH(`{field}`) = %s AND DAY(`{field}`) = %s
               AND docstatus < 2""".format(
                doctype=rule["document_type"],
                field=rule["birthdate_field"]
            ),
            (target.month, target.day)
        )
        return int(result[0][0]) if result else 0
    except Exception:
        return 0


def _birthday_channels(rule):
    channels = []
    if rule.get("send_to_person"):
        channels.append("Person")
    if rule.get("send_to_group"):
        channels.append("Group")
    if rule.get("send_to_additional"):
        channels.append("Additional")
    return ", ".join(channels) if channels else "-"


# ---------------------------------------------------------------------------
# Days Before / After Notification Rules
# ---------------------------------------------------------------------------

def _get_date_event_rules():
    """Return one entry per rule (grouped), each with a nested `offsets` list."""
    rules = frappe.get_all(
        "WhatsApp Notification Rule",
        filters={"event": ["in", ["Days Before", "Days After"]]},
        fields=[
            "name", "rule_name", "document_type", "event", "days_offset",
            "date_field", "enabled", "enable_active_hours",
            "active_hours_start", "active_hours_end", "send_once"
        ],
        order_by="enabled desc, document_type asc"
    )

    today_str = today()
    result = []

    for rule in rules:
        reminder_rows = frappe.get_all(
            "WhatsApp Notification Reminder",
            filters={"parent": rule["name"], "parenttype": "WhatsApp Notification Rule"},
            fields=["days"],
            order_by="days desc"
        )

        offset_entries = []
        if reminder_rows:
            rule["has_reminder_offsets"] = True
            for row in reminder_rows:
                days = int(row["days"] or 0)
                if not days:
                    continue
                target = _calc_target_date_from_days(rule["event"], today_str, days)
                match_info = _get_date_rule_matches(rule, target)
                sent = _count_sent_today_offset(rule["name"], days)
                offset_entries.append({
                    "days": days,
                    "days_left_label": _days_left_label(rule["event"], days),
                    "target_date": str(target) if target else None,
                    "matches_today": match_info["count"],
                    "sent_today": sent,
                    "is_done": match_info["count"] > 0 and sent >= match_info["count"],
                    "matching_docs": match_info["docs"],
                    "has_more": match_info["has_more"],
                })
        else:
            rule["has_reminder_offsets"] = False
            days = int(rule.get("days_offset") or 0)
            target = _calc_target_date(rule, today_str)
            match_info = _get_date_rule_matches(rule, target)
            sent = _count_sent_today(rule["name"])
            offset_entries.append({
                "days": days,
                "days_left_label": _days_left_label(rule["event"], days),
                "target_date": str(target) if target else None,
                "matches_today": match_info["count"],
                "sent_today": sent,
                "is_done": match_info["count"] > 0 and sent >= match_info["count"],
                "matching_docs": match_info["docs"],
                "has_more": match_info["has_more"],
            })

        rule["offsets"] = offset_entries
        rule["total_matches_today"] = sum(o["matches_today"] for o in offset_entries)
        rule["total_sent_today"] = sum(o["sent_today"] for o in offset_entries)
        rule["last_sent"] = _get_last_sent(rule["name"])
        rule["active_hours_label"] = _active_hours_label(rule)
        result.append(rule)

    return result


def _calc_target_date(rule, today_str):
    if not rule.get("days_offset"):
        return None
    offset = int(rule["days_offset"])
    if rule["event"] == "Days Before":
        return getdate(add_days(today_str, offset))
    else:  # Days After
        return getdate(add_days(today_str, -offset))


def _calc_target_date_from_days(event, today_str, days):
    if not days:
        return None
    if event == "Days Before":
        return getdate(add_days(today_str, days))
    else:
        return getdate(add_days(today_str, -days))


def _get_date_rule_matches(rule, target_date, limit=20):
    if not rule.get("document_type") or not rule.get("date_field") or not target_date:
        return {"count": 0, "docs": [], "has_more": False}
    try:
        all_names = frappe.get_all(
            rule["document_type"],
            filters={rule["date_field"]: str(target_date)},
            pluck="name"
        )
        count = len(all_names)
        docs = [{"name": n, "doctype": rule["document_type"]} for n in all_names[:limit]]
        return {"count": count, "docs": docs, "has_more": count > limit}
    except Exception:
        return {"count": 0, "docs": [], "has_more": False}


def _count_sent_today(rule_name):
    try:
        result = frappe.db.sql(
            """SELECT COUNT(*) FROM `tabWhatsApp Message Log`
               WHERE notification_rule = %s
               AND DATE(creation) = CURDATE()""",
            rule_name
        )
        return int(result[0][0]) if result else 0
    except Exception:
        return 0


def _count_sent_today_offset(rule_name, offset_days):
    try:
        result = frappe.db.sql(
            """SELECT COUNT(*) FROM `tabWhatsApp Message Log`
               WHERE notification_rule = %s
               AND reminder_days_offset = %s
               AND DATE(creation) = CURDATE()""",
            (rule_name, offset_days)
        )
        return int(result[0][0]) if result else 0
    except Exception:
        return 0


def _get_last_sent(rule_name):
    try:
        result = frappe.db.sql(
            """SELECT creation FROM `tabWhatsApp Message Log`
               WHERE notification_rule = %s
               ORDER BY creation DESC LIMIT 1""",
            rule_name,
            as_dict=True
        )
        return str(result[0]["creation"]) if result else None
    except Exception:
        return None


def _active_hours_label(rule):
    if not rule.get("enable_active_hours"):
        return "Anytime"
    start = (rule.get("active_hours_start") or "")[:5]
    end = (rule.get("active_hours_end") or "")[:5]
    if start and end:
        return "{} – {}".format(start, end)
    return "Restricted"


def _days_left_label(event, days):
    """Human-readable label for an offset, e.g. 'In 14 days', 'Tomorrow', 'Today'."""
    if days == 0:
        return "Today"
    if event == "Days Before":
        if days == 1:
            return "Tomorrow"
        return "In {} days".format(days)
    else:  # Days After
        if days == 1:
            return "Yesterday"
        return "{} days ago".format(days)


# ---------------------------------------------------------------------------
# Recent Activity
# ---------------------------------------------------------------------------

@frappe.whitelist()
def get_bulk_sends(limit=20):
    """Return recent Envio em Massa WhatsApp docs for the monitor page.

    Always shows non-completed sends; completed sends are only shown if created
    within the last 48 hours to keep the monitor view clean.
    """
    try:
        from frappe.utils import now_datetime, add_days
        cutoff = str(add_days(today(), -2))
        envios = frappe.db.sql("""
            SELECT name, aviso, titulo, status,
                   total, enviados, falhados, pendentes,
                   disparado_por, iniciado_em, concluido_em, ultimo_heartbeat
            FROM `tabEnvio em Massa WhatsApp`
            WHERE status != 'Concluído'
               OR DATE(creation) >= %s
            ORDER BY creation DESC
            LIMIT %s
        """, (cutoff, int(limit)), as_dict=True)

        now = str(now_datetime())
        for e in envios:
            done = (e.get("enviados") or 0) + (e.get("falhados") or 0)
            total = e.get("total") or 1
            e["progresso"] = round(done / total * 100)
            if e.get("status") == "Em Execução" and e.get("ultimo_heartbeat"):
                from frappe.utils import time_diff_in_seconds
                diff = time_diff_in_seconds(now, str(e["ultimo_heartbeat"]))
                if diff > 900:
                    e["possivelmente_interrompido"] = True
        return envios
    except Exception:
        frappe.log_error(title="get_bulk_sends error", message=frappe.get_traceback())
        return []


def _get_recent_activity():
    # Collect names of all date-event notification rules
    date_rule_names = frappe.get_all(
        "WhatsApp Notification Rule",
        filters={"event": ["in", ["Days Before", "Days After"]]},
        pluck="name"
    )

    if not date_rule_names:
        return []

    try:
        logs = frappe.get_all(
            "WhatsApp Message Log",
            filters={"notification_rule": ["in", date_rule_names]},
            fields=[
                "name", "notification_rule", "reference_doctype", "reference_name",
                "recipient_name", "phone", "status", "creation", "error_message"
            ],
            order_by="creation desc",
            limit=30
        )
        return logs
    except Exception:
        return []

File: whatsapp_notifications/whatsapp_notifications/page/whatsapp_schedule_monitor/whatsapp_schedule_monitor.js
frappe.pages["whatsapp-schedule-monitor"].on_page_load = function (wrapper) {
    var page = frappe.ui.make_app_page({
        parent: wrapper,
        title: "WhatsApp Schedule Monitor",
        single_column: true,
    });

    page.add_action_item(__("Refresh"), () => load_data(page));

    frappe.breadcrumbs.add("WhatsApp Notifications");

    $(wrapper).find(".layout-main-section").html(render_skeleton());

    load_data(page);
};

function load_data(page) {
    frappe.call({
        method: "whatsapp_notifications.whatsapp_notifications.page.whatsapp_schedule_monitor.whatsapp_schedule_monitor.get_schedule_data",
        callback: function (r) {
            if (r.message) {
                render_page(page, r.message);
            }
        },
        error: function () {
            frappe.msgprint(__("Error loading schedule data."));
        }
    });
}

// ---------------------------------------------------------------------------
// Render helpers
// ---------------------------------------------------------------------------

function render_skeleton() {
    return `<div class="schedule-monitor-container" style="padding:20px">
        <p class="text-muted">${__("Loading...")} </p>
    </div>`;
}

function render_page(page, data) {
    var now = frappe.datetime.str_to_user(data.now);
    var html = `
        <div class="schedule-monitor-container" style="padding:4px 0">
            <div class="text-muted" style="margin-bottom:18px;font-size:12px;">
                ${__("Last refreshed")}: <b>${now}</b>
                &nbsp;|&nbsp; ${__("Today")}: <b>${frappe.datetime.str_to_user(data.today)}</b>
            </div>
            ${render_bulk_sends_section(data.bulk_sends || [])}
            <div style="margin-top:28px"></div>
            ${render_birthday_section(data.birthday_rules)}
            <div style="margin-top:28px"></div>
            ${render_date_event_section(data.date_event_rules)}
            <div style="margin-top:28px"></div>
            ${render_recent_activity(data.recent_activity)}
        </div>`;

    $(page.wrapper).find(".layout-main-section").html(html);

    // Wire run-now buttons
    $(page.wrapper).find(".btn-run-birthday").on("click", function () {
        var rule = $(this).data("rule");
        run_birthday_now(rule, page);
    });

    // Wire retomar buttons
    $(page.wrapper).find(".btn-retomar-envio").on("click", function () {
        var envio_name = $(this).data("envio");
        retomar_envio(envio_name, page);
    });

    // Auto-refresh if any envio is running
    var has_running = (data.bulk_sends || []).some(e => e.status === 'Em Execução');
    if (has_running) {
        if (!page._bulk_refresh_timer) {
            page._bulk_refresh_timer = setInterval(() => load_data(page), 8000);
        }
    } else {
        if (page._bulk_refresh_timer) {
            clearInterval(page._bulk_refresh_timer);
            page._bulk_refresh_timer = null;
        }
    }
}

// ---------------------------------------------------------------------------
// Birthday Rules section
// ---------------------------------------------------------------------------

function render_birthday_section(rules) {
    var rows = (rules || []).map(r => render_birthday_row(r)).join("");

    return `
    <div class="card" style="${card_style()}">
        <div style="${section_header_style()}">
            <span style="font-size:16px;font-weight:600;">
                🎂 ${__("Birthday Rules")}
            </span>
            <span class="badge" style="background:#f0f4f8;color:#6c757d;padding:4px 10px;border-radius:20px;font-size:12px;">
                ${rules.length} ${__("rules")}
            </span>
        </div>
        ${rules.length === 0
            ? `<p class="text-muted" style="padding:16px;">${__("No birthday rules configured.")}</p>`
            : `<div style="overflow-x:auto;">
                <table class="table table-sm" style="${table_style()}">
                    <thead>
                        <tr style="background:#f8f9fa;">
                            <th>${__("Rule")}</th>
                            <th>${__("DocType")}</th>
                            <th>${__("Days Before")}</th>
                            <th>${__("Send Time")}</th>
                            <th>${__("Channels")}</th>
                            <th>${__("Status")}</th>
                            <th>${__("Today's Matches")}</th>
                            <th>${__("Next Run")}</th>
                            <th>${__("Last Run")}</th>
                            <th>${__("Last Result")}</th>
                            <th></th>
                        </tr>
                    </thead>
                    <tbody>${rows}</tbody>
                </table>
               </div>`
        }
    </div>`;
}

function render_birthday_row(r) {
    var enabled_badge = r.enabled
        ? `<span style="${badge_style("green")}">${__("Active")}</span>`
        : `<span style="${badge_style("gray")}">${__("Disabled")}</span>`;

    var next_run_html = "";
    if (!r.enabled) {
        next_run_html = `<span class="text-muted">—</span>`;
    } else if (r.next_run === "overdue") {
        next_run_html = `<span style="${badge_style("orange")}">${__("Overdue")}</span>`;
    } else if (r.next_run) {
        next_run_html = frappe.datetime.str_to_user(r.next_run);
    } else {
        next_run_html = `<span class="text-muted">${__("No send time set")}</span>`;
    }

    var last_run_html = r.last_run
        ? frappe.datetime.str_to_user(r.last_run)
        : `<span class="text-muted">—</span>`;

    var last_result_html = r.last_status
        ? `<span title="${r.last_error || ""}">${r.last_status}${r.last_count ? ` (${r.last_count})` : ""}</span>`
        : `<span class="text-muted">—</span>`;

    var matches_html = `<b>${r.matches_today}</b>`;
    if (r.was_run_today) {
        matches_html += ` <span style="${badge_style("blue")}">${__("Sent")}</span>`;
    }

    var days_label = r.days_before == 0
        ? __("On the day")
        : `${r.days_before} ${__("day(s) before")}`;

    return `<tr>
        <td><a href="/app/whatsapp-birthday-rule/${r.name}">${r.name}</a></td>
        <td>${r.document_type || "—"}</td>
        <td>${days_label}</td>
        <td><b>${r.send_time || "—"}</b></td>
        <td style="font-size:11px;">${r.send_channels || "—"}</td>
        <td>${enabled_badge}</td>
        <td style="text-align:center;">${matches_html}</td>
        <td>${next_run_html}</td>
        <td style="font-size:12px;">${last_run_html}</td>
        <td style="font-size:12px;">${last_result_html}</td>
        <td>
            <button class="btn btn-xs btn-default btn-run-birthday" data-rule="${r.name}">
                ${__("Run Now")}
            </button>
        </td>
    </tr>`;
}

function run_birthday_now(rule_name, page) {
    frappe.confirm(
        __("Run birthday rule <b>{0}</b> now?", [rule_name]),
        function () {
            frappe.call({
                method: "whatsapp_notifications.whatsapp_notifications.doctype.whatsapp_birthday_rule.whatsapp_birthday_rule.run_now",
                args: { doctype: "WhatsApp Birthday Rule", docname: rule_name, force: 1 },
                callback: function (r) {
                    if (r.message) {
                        var m = r.message;
                        frappe.msgprint(
                            __("Sent: {0} | Skipped: {1} | Errors: {2}", [m.sent, m.skipped, m.errors.length])
                        );
                        load_data(page);
                    }
                }
            });
        }
    );
}

// ---------------------------------------------------------------------------
// Days Before / After Rules section
// ---------------------------------------------------------------------------

function render_date_event_section(rules) {
    var rows = (rules || []).map(r => render_date_event_row(r)).join("");

    return `
    <div class="card" style="${card_style()}">
        <div style="${section_header_style()}">
            <span style="font-size:16px;font-weight:600;">
                📅 ${__("Days Before / After Rules")}
            </span>
            <span class="badge" style="background:#f0f4f8;color:#6c757d;padding:4px 10px;border-radius:20px;font-size:12px;">
                ${rules.length} ${__("rules")}
            </span>
        </div>
        ${rules.length === 0
            ? `<p class="text-muted" style="padding:16px;">${__("No Days Before / After rules configured.")}</p>`
            : `<div style="overflow-x:auto;">
                <table class="table table-sm" style="${table_style()}">
                    <thead>
                        <tr style="background:#f8f9fa;">
                            <th>${__("Rule Name")}</th>
                            <th>${__("DocType")}</th>
                            <th>${__("Event")}</th>
                            <th>${__("Date Field")}</th>
                            <th>${__("Target Date")}</th>
                            <th>${__("Status")}</th>
                            <th>${__("Matches Today")}</th>
                            <th>${__("Sent Today")}</th>
                            <th>${__("Active Hours")}</th>
                            <th>${__("Last Sent")}</th>
                        </tr>
                    </thead>
                    <tbody>${rows}</tbody>
                </table>
               </div>`
        }
    </div>`;
}

function render_date_event_row(r) {
    var enabled_badge = r.enabled
        ? `<span style="${badge_style("green")}">${__("Active")}</span>`
        : `<span style="${badge_style("gray")}">${__("Disabled")}</span>`;

    var event_label = r.event === "Days Before"
        ? `<span style="color:#e65100;">${r.days_offset} ${__("day(s) before")}</span>`
        : `<span style="color:#1565c0;">${r.days_offset} ${__("day(s) after")}</span>`;

    var matches_html = (r.matches_today || 0).toString();
    if (r.matches_today > 0 && r.sent_today === 0) {
        matches_html = `<b style="color:#e65100;">${r.matches_today}</b>
            <span style="${badge_style("orange")}">${__("Pending")}</span>`;
    } else if (r.matches_today > 0 && r.sent_today > 0) {
        matches_html = `<b>${r.matches_today}</b>`;
    }

    var sent_today_html = r.sent_today > 0
        ? `<b style="color:#2e7d32;">${r.sent_today}</b>`
        : `<span class="text-muted">0</span>`;

    var last_sent_html = r.last_sent
        ? `<span style="font-size:11px;">${frappe.datetime.str_to_user(r.last_sent)}</span>`
        : `<span class="text-muted">—</span>`;

    return `<tr>
        <td><a href="/app/whatsapp-notification-rule/${r.name}">${r.rule_name || r.name}</a></td>
        <td>${r.document_type || "—"}</td>
        <td>${event_label}</td>
        <td style="font-size:12px;">${r.date_field || "—"}</td>
        <td style="font-size:12px;"><b>${r.target_date || "—"}</b></td>
        <td>${enabled_badge}</td>
        <td style="text-align:center;">${matches_html}</td>
        <td style="text-align:center;">${sent_today_html}</td>
        <td style="font-size:12px;">${r.active_hours_label || "Anytime"}</td>
        <td>${last_sent_html}</td>
    </tr>`;
}

// ---------------------------------------------------------------------------
// Recent Activity
// ---------------------------------------------------------------------------

function render_recent_activity(logs) {
    if (!logs || logs.length === 0) {
        return `
        <div class="card" style="${card_style()}">
            <div style="${section_header_style()}">
                <span style="font-size:16px;font-weight:600;">📨 ${__("Recent Activity")}</span>
            </div>
            <p class="text-muted" style="padding:16px;">${__("No recent activity from scheduled rules.")}</p>
        </div>`;
    }

    var rows = logs.map(function (log) {
        var status_color = {
            "Sent": "green", "Failed": "red", "Pending": "orange",
            "Queued": "blue", "Sending": "blue"
        }[log.status] || "gray";

        var status_badge = `<span style="${badge_style(status_color)}">${log.status}</span>`;

        var rule_link = log.notification_rule
            ? `<a href="/app/whatsapp-notification-rule/${log.notification_rule}" style="font-size:11px;">${log.notification_rule}</a>`
            : "—";

        var doc_link = log.reference_doctype && log.reference_name
            ? `<a href="/app/${frappe.router.slug(log.reference_doctype)}/${log.reference_name}" style="font-size:11px;">${log.reference_name}</a>`
            : "—";

        var error_html = log.error_message
            ? `<span class="text-danger" title="${log.error_message}" style="font-size:10px;cursor:help;">⚠</span>`
            : "";

        return `<tr>
            <td style="font-size:11px;">${frappe.datetime.str_to_user(log.creation)}</td>
            <td>${rule_link}</td>
            <td style="font-size:11px;">${log.reference_doctype || "—"}</td>
            <td>${doc_link}</td>
            <td style="font-size:12px;">${log.recipient_name || log.phone || "—"}</td>
            <td>${status_badge} ${error_html}</td>
        </tr>`;
    }).join("");

    return `
    <div class="card" style="${card_style()}">
        <div style="${section_header_style()}">
            <span style="font-size:16px;font-weight:600;">📨 ${__("Recent Activity")}</span>
            <span class="badge" style="background:#f0f4f8;color:#6c757d;padding:4px 10px;border-radius:20px;font-size:12px;">
                ${__("Last")} ${logs.length} ${__("messages")}
            </span>
        </div>
        <div style="overflow-x:auto;">
            <table class="table table-sm" style="${table_style()}">
                <thead>
                    <tr style="background:#f8f9fa;">
                        <th>${__("Time")}</th>
                        <th>${__("Rule")}</th>
                        <th>${__("DocType")}</th>
                        <th>${__("Document")}</th>
                        <th>${__("Recipient")}</th>
                        <th>${__("Status")}</th>
                    </tr>
                </thead>
                <tbody>${rows}</tbody>
            </table>
        </div>
    </div>`;
}

// ---------------------------------------------------------------------------
// Bulk Sends section
// ---------------------------------------------------------------------------

function render_bulk_sends_section(envios) {
    var rows = (envios || []).map(e => render_bulk_send_row(e)).join("");

    return `
    <div class="card" style="${card_style()}">
        <div style="${section_header_style()}">
            <span style="font-size:16px;font-weight:600;">
                📤 ${__("Envios em Massa")}
            </span>
            <span class="badge" style="background:#f0f4f8;color:#6c757d;padding:4px 10px;border-radius:20px;font-size:12px;">
                ${envios.length} ${__("envios")}
            </span>
        </div>
        ${envios.length === 0
            ? `<p class="text-muted" style="padding:16px;">${__("Nenhum envio em massa encontrado.")}</p>`
            : `<div style="overflow-x:auto;">
                <table class="table table-sm" style="${table_style()}">
                    <thead>
                        <tr style="background:#f8f9fa;">
                            <th>${__("Envio")}</th>
                            <th>${__("Aviso")}</th>
                            <th>${__("Status")}</th>
                            <th>${__("Progresso")}</th>
                            <th>${__("Total")}</th>
                            <th>${__("Enviados")}</th>
                            <th>${__("Falhados")}</th>
                            <th>${__("Disparado Por")}</th>
                            <th>${__("Iniciado Em")}</th>
                            <th>${__("Concluído Em")}</th>
                            <th></th>
                        </tr>
                    </thead>
                    <tbody>${rows}</tbody>
                </table>
               </div>`
        }
    </div>`;
}

function render_bulk_send_row(e) {
    var status_colors = {
        "Preparando": "blue", "Em Execução": "blue",
        "Concluído": "green", "Interrompido": "orange", "Falhado": "red"
    };
    var status_badge = `<span style="${badge_style(status_colors[e.status] || 'gray')}">${e.status}</span>`;
    if (e.possivelmente_interrompido) {
        status_badge += ` <span style="${badge_style('orange')}" title="${__('Sem heartbeat há mais de 15 min')}">⚠</span>`;
    }

    var pct = e.progresso || 0;
    var bar_color = e.status === 'Concluído' ? '#4caf50' : (e.status === 'Em Execução' ? '#2196f3' : (e.falhados > 0 ? '#ff9800' : '#9e9e9e'));
    var progress_html = `
        <div style="min-width:120px;">
            <div style="font-size:11px;margin-bottom:2px;">${pct}% (${(e.enviados||0)+(e.falhados||0)}/${e.total||0})</div>
            <div style="height:6px;background:#e0e0e0;border-radius:3px;">
                <div style="height:100%;width:${pct}%;background:${bar_color};border-radius:3px;"></div>
            </div>
        </div>`;

    var aviso_link = e.aviso
        ? `<a href="/app/aviso-whatsapp/${e.aviso}" style="font-size:12px;">${e.titulo || e.aviso}</a>`
        : '—';

    var iniciado = e.iniciado_em ? `<span style="font-size:11px;">${frappe.datetime.str_to_user(e.iniciado_em)}</span>` : '—';
    var concluido = e.concluido_em ? `<span style="font-size:11px;">${frappe.datetime.str_to_user(e.concluido_em)}</span>` : '—';

    var can_resume = ['Em Execução', 'Interrompido', 'Falhado'].includes(e.status) && (e.pendentes || 0) > 0;
    var action_btn = can_resume
        ? `<button class="btn btn-xs btn-warning btn-retomar-envio" data-envio="${e.name}">↺ ${__("Retomar")}</button>`
        : `<a href="/app/envio-em-massa-whatsapp/${e.name}" class="btn btn-xs btn-default">${__("Ver")}</a>`;

    return `<tr>
        <td><a href="/app/envio-em-massa-whatsapp/${e.name}" style="font-size:12px;">${e.name}</a></td>
        <td>${aviso_link}</td>
        <td>${status_badge}</td>
        <td>${progress_html}</td>
        <td style="text-align:center;">${e.total || 0}</td>
        <td style="text-align:center;color:#2e7d32;"><b>${e.enviados || 0}</b></td>
        <td style="text-align:center;color:#c62828;">${e.falhados || 0}</td>
        <td style="font-size:11px;">${e.disparado_por || '—'}</td>
        <td>${iniciado}</td>
        <td>${concluido}</td>
        <td>${action_btn}</td>
    </tr>`;
}

function retomar_envio(envio_name, page) {
    frappe.confirm(
        __("Retomar o envio <b>{0}</b>?", [envio_name]),
        function () {
            frappe.call({
                method: "whatsapp_notifications.whatsapp_notifications.doctype.envio_em_massa_whatsapp.envio_em_massa_whatsapp.retomar_envio",
                args: { envio_name: envio_name },
                callback: function(r) {
                    if (r.message && r.message.queued) {
                        frappe.show_alert({ message: __("Envio retomado! {0} pendentes.", [r.message.pendentes]), indicator: "green" });
                        setTimeout(() => load_data(page), 2000);
                    }
                }
            });
        }
    );
}

// ---------------------------------------------------------------------------
// Style helpers
// ---------------------------------------------------------------------------

function card_style() {
    return "border:1px solid #e0e0e0;border-radius:8px;background:#fff;overflow:hidden;";
}

function section_header_style() {
    return "display:flex;align-items:center;justify-content:space-between;padding:14px 18px;border-bottom:1px solid #f0f0f0;";
}

function table_style() {
    return "margin:0;font-size:13px;";
}

function badge_style(color) {
    var colors = {
        green:  "background:#e8f5e9;color:#2e7d32;",
        red:    "background:#ffebee;color:#c62828;",
        orange: "background:#fff3e0;color:#e65100;",
        blue:   "background:#e3f2fd;color:#1565c0;",
        gray:   "background:#f5f5f5;color:#757575;",
    };
    return `display:inline-block;padding:2px 8px;border-radius:10px;font-size:11px;font-weight:500;${colors[color] || colors.gray}`;
}
